pkg/filter/filter.go

```diff
@@ -0,0 +1,67 @@
+package filter
+
+import (
+	"net/http"
+)
+
+// Request contains the original http.Request  and some additional data.
+type Request struct {
+	// Request is the original http.Request
+	*http.Request
+
+	// PathParams contains the URL path parameters
+	PathParams map[string]string
+
+	// Body is the loaded request body (usually JSON)
+	Body []byte
+}
+
+// Response defines the attributes of the HTTP response that will be sent to the client
+type Response struct {
+	// StatusCode is the HTTP status code
+	StatusCode int
+
+	// Header contains the response headers
+	Header http.Header
+
+	// Body is the response body (usually JSON)
+	Body []byte
+}
+
+// Handler is a function that handles a request after all the applicable filters
+type Handler func(*Request) (*Response, error)
+
+// Middleware is a function that intercepts a request before it reaches the final handler.
+// Normally this function should invoke next with the request to proceed with
+// the next middleware or the final handler.
+// It can also terminate the request processing by not invoking next.
+type Middleware func(req *Request, next Handler) (*Response, error)
+
+// RouteMatcher defines the criteria to match against registered routes.
+//
+// NOTE: This does not match against HTTP requests at runtime but registered
+// routes at startup. This improves performance but has some limitations.
+type RouteMatcher struct {
+	// Methods match route method.
+	// If nil, matches any method.
```

Comment:
Not sure if nil is good match. I'd rather have "*" as a match because then you explicitly state you want all methods. Otherwise you could've forgot to put any methods altogether